sdk/tracing/src/main/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessor.java

```diff
@@ -217,25 +218,24 @@ public void run() {
         // avoid blocking the producer thread.
         ArrayList<ReadableSpan> spansCopy;
         synchronized (monitor) {
-          // If still maxExportBatchSize elements in the queue better to execute an extra
-          if (spansList.size() < maxExportBatchSize) {
-            do {
-              // In the case of a spurious wakeup we export only if we have at least one span in
-              // the batch. It is acceptable because batching is a best effort mechanism here.
-              try {
-                monitor.wait(scheduleDelayMillis);
-              } catch (InterruptedException ie) {
-                // Preserve the interruption status as per guidance and stop doing any work.
-                Thread.currentThread().interrupt();
-                return;
-              }
-            } while (spansList.isEmpty());
-          }
+          do {
+            // In the case of a spurious wakeup we export only if we have at least one span in
+            // the batch. It is acceptable because batching is a best effort mechanism here.
+            try {
+              monitor.wait(scheduleDelayMillis);
```

Comment:
I think this causes the wait to happen between each export batch, rather than between each "processing of the queue". It effectively means the queue size parameter is not having much effect if I'm not mistaken. 

FWIW, I am interested in flushing in an async manner :) I think the way to do it is to have a separate Flusher. What I'm expecting is for it to look like something like

```java
ExecutorService executor;

init() executor.schedule(wakeup);

wakeup {
  if (spans.empty)
    executor.schedule(wakeup)

  drained = spans.drain

  new Flusher(drained).flush();

Flusher {
  pending

  Flusher(spans) pending = spans
  
  flush() {
    if (pending.isEmpty()) {
      executor.schedule(wakeup)
      return
    }

    taken = pending.take(batchSize)
    result = exporter.export(taken)
    result.whenComplete(this::flush)
}
```

I think this has the properties that we want - run through entire queue when exporting, export a batch at a time, and delay between each queue processing (not between every batch as in the right side here, nor potentially no delay ever like left side here). And consecutive flushes without blocking.

So my plan is #1565, then #729, then add AsyncFlusher. @jkwatson has wanted us to migrate to an `ExecutorService` for a long time and I agree :) I think it will make our lives simpler when actually addressing the asyncronous export. Does this plan make sense?

Reply:
@anuraaga I need to absorb your comments above more and will do so now. However, #1565 does not make sense to me as it introduces additional complexity that is avoidable. Back soon.

Reply:
So, if I've got this right... you're making a distinction here between the time interval in terms of processing the queue vs the time interval between exporting to an exporter. You're also favouring the former.

You want to see exporting happen continually if there are more elements in the queue.

Correct?

Reply:
Yeah - at least that's my understanding of the spec / goal of the class, but I could be wrong in my reading of course :)

https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/sdk.md#batching-processor

A second reading makes me think actually this PR may have the right semantics since it says scheduledDelayMillis = delay between export, and maxExportBatchSize = maximum size of an export. I'm having trouble invisioning how I would actually set maxExportBatchSize different from queueSize but this PR seems ok to me too now. Interested in hearing what others think the BSP is "supposed to do" though :)

Reply:
OK, I've just pushed another commit.

The worker thread has been removed and we now instead use the timer for scheduling work, thus saving us a thread. In addition, if there remains work to be done once the export has completed then we queue it for running immediately.

A test has changed slightly in that it was previously relying on the worker thread becoming blocked; which it now no longer does.